File: CS2336/Lab00/Python/Solution/Statistician.py
#!/usr/bin/env python3
'''This class models a simple Statistician.'''

from sys import float_info
from typing import TypeVar
from math import sqrt

Statistician = TypeVar('Statistician')
Number = TypeVar('Number', int, float)


class Statistician(object):
    ''' A Statistician manages a stream of numbers and provides various
        statistics: mean, standard deviation, smallest, largest, etc.
    '''
    # Class invariant:
    #   The instance variables __N, __sumX, __sumXsq, __smallest,
    #   and __largest are used to maintain state of a
    #   Statistician object.  __N and __sumX are used to
    #   calculate the mean.  These attributes plus __sumXsq are used
    #   to calculate the standard deviation

    def __init__(self, other:Statistician=None) -> Statistician:
        '''
        Statistician class constructor front end

        Postcondition:
            creates a newly constructed, well-defined Statistician
            object -- initialized as though it has seen no
            numbers OR creates a copy of other
        '''
        if other is None:
            self.__default_constructor()
        else:
            if not isinstance(other, Statistician):
                raise TypeError("other not an instance of Statistician")
            self.__copy_constructor(other)
    # end __init__

    def __default_constructor(self) -> Statistician:
        '''
        Statistician class constructor.

        Postcondition:
            a newly constructed, well-defined Statistician
            object -- initialized as though it has seen no
            numbers
        '''
        self.__N       = 0                 # How many numbers in the sequence
        self.__sumX    = 0.0               # Sum of all the numbers
        self.__sumXsq  = 0.0               # Sum of all numbers squared
        self.__smallest = float_info.max   # The smallest number in the sequenc
        self.__largest = -float_info.max   # The largest number in the sequence
    # end __default_constructor

    def __copy_constructor(self, other:Statistician) -> Statistician:
        '''Statistician class copy constructor.

        Argument:
            other (a reference to a Statistician instance)

        Postcondition:
            a newly constructed Statistician that
            whose attributes are a copy of the
            attributes of other

        Raises:
            TypeError if other not reference to instance of Statistician
        '''
        if not isinstance(other, Statistician):
            raise TypeError("other not an instance of Statistician")
        self.__N        = other.__N
        self.__sumX     = other.__sumX
        self.__sumXsq   = other.__sumXsq
        self.__smallest = other.__smallest
        self.__largest  = other.__largest
    # end __copy_constructor

    def __eq__(self, other:Statistician) -> bool:
        '''
        Determins whether this instance is equal in value
           to other.

        Argument:
               other (a reference to a Statistician instance)

        Returns:
               True if other an instance of Statistician and other's
               attributes are identical to the attributes of this object

               False otherwise
        '''
        if not isinstance(other, Statistician):
            return False

        if self is other:  # are we comparing with ourself
            return True

        if not hash(self) == hash(other):
            return False

        return ( (self.__N == other.__N) and
                 (self.__sumX == other.__sumX) and
                 (self.__sumXsq == other.__sumXsq) and
                 (self.__smallest == other.__smallest) and
                 (self.__largest == other.__largest)
               )
    # end __eq__

    def __ne__(self, other:Statistician) -> bool:
        '''The negation of __eq__.'''
        return not (self == other)
    # end __ne__

    def __hash__(self) -> int:
        '''
        Calculates the hash value of this object

        Preconditions: none

        Returns: an integer representing the combined hash
                 value of all this object's attributes.
        '''
        return hash((self.__N, self.__sumX, self.__sumXsq, self.__smallest,
                    self.__largest))
    # end __hash__

    def length(self) -> int:
        '''
        Determine how many numbers have been given to this Statistician

        Returns:
            the count of how many numbers have been given to this
            Statistician since it was initialized or reinitialized
        '''
        return self.__N
    # end length

    def largest(self) -> float:
        '''
        Determine the largest number that has been given to this
        Statistician

        Returns:
            the largest number that has been given to this
            Statistician since it was initialized or reinitialized

            NOTE: if length() is 0, then returns float('nan')
        '''
        if self.length() == 0:
            return float('nan')
        else:
            return self.__largest

    def mean(self) -> float:
        '''
        Determine the arithmetic average of all the numbers
           given to this Statistician.

        Returns:
            the arithmetic average/mean of all the numbers given
            to this Statistician since it was initialized or reinitialized

            NOTE: if length() is 0, then returns float('nan')
        '''
        if self.length() == 0:
            return float('nan')
        else:
            # STUDENT complete this ensuring that result returned is float
            #         replace 0.0 with the appropriate expression
            return 0.0
    # end mean

    def next_number(self, number:Number) -> None:
        '''
        Give a new number to this Statistician.

        Argument:
            number must be a number

        Precondition:
            number must be a float number

        Postcondition:
            The specified number has been given to this Statistician
            and it will be included in any subsequent statistics

        Raises:
            TypeError if number not an instance of a number
        '''
        if not (isinstance(number, float) or isinstance(number, int)):
            raise TypeError("argument is not a number")

        # STUDENT completes implementation here.
        #   NOTE: if number is the first number given to the statistician,
        #         then number is both the largest and smallest seen so far.

    # end nextNumber

    def reset(self) -> None:
        '''
        Reset this Statistician to the initial state with no numbers
           in the sequence.

        Postcondition:
            This Statistician is reinitialized as if it has never been
            given any numbers
        '''
        self.__N       = 0                 # How many numbers in the sequence
        self.__sumX    = 0.0               # Sum of all the numbers
        self.__sumXsq  = 0.0               # Sum of all the squares of numbers
        self.__smallest = float_info.max   # The smallest number in the sequence
        self.__largest = -float_info.max   # The largest number in the sequence
    # end reset

    @staticmethod
    def scale(scale:Number, statistician:Statistician) -> Statistician:
        '''
        Create a new Statistician whose state is scaled by parameter
           scale -- such that statistics (other than length() are
           multiplied by scale.

        Arguments:
            scale -- must be a number
            statistician -- must be an instance of Statistician

        Raises:
            TypeError if either scale or statistician not of
            appropriate type

        Returns:
            a new Statistician that is a scaled version of statistician
        '''
        if not (isinstance(scale, float) or isinstance(scale, int)):
            raise TypeError("scale argument is not a float number")
        if not isinstance(statistician, Statistician):
            raise TypeError("statistician argument is not a Statistician")

        temp = Statistician(statistician)

        # STUDENT complete implementation here

        return temp
    # end scale

    def smallest(self) -> float:
        '''
        Determine the smallest number that has been given to this
           Statistician.

        Returns:
            the smallest number that has been given to this
            Statistician since it was initialized or reinitialized

            NOTE: if length() is 0, then returns float('nan')
        '''
        if self.length() == 0:
            return float('nan')
        else:
            return self.__smallest
    # end smallest

    def stdDev(self) -> float:
        '''
        Determine the standard deviation of the stream of values
           given to this statistician.

        Returns:
            The calculated standard deviation for a population of values
        '''
        # STUDENT complete this by replacing the pass with appropriate
        #         return
        pass
    # end stdDev

    def __str__(self) -> str:
        '''
        Create a string representation of this Statistician.

        Returns:
            a string representation of this Statistician

        NOTE:  Since strings are immutable, please note the method
               used
        '''
        buffer = []
        buffer.append("Statistician: \n")
        buffer.append("\tN = " + str(self.__N) + "\n")
        buffer.append("\tsumX = " + str(self.__sumX) + "\n")
        buffer.append("\tsumXsq = " + str(self.__sumXsq) + "\n")
        if self.__N == 0:
            buffer.append("\tmean    = NOT DEFINED\n")
            buffer.append("\tstd-dev = NOT DEFINED\n")
            buffer.append("\tmin     = NOT DEFINED\n")
            buffer.append("\tmax     = NOT DEFINED\n")
        else:
            buffer.append("\tmean     = " + str(self.mean()) + "\n")
            buffer.append("\tstd-dev  = " + str(self.stdDev()) + "\n")
            buffer.append("\tmin      = " + str(self.__smallest) + "\n")
            buffer.append("\tmax      = " + str(self.__largest) + "\n")
        # end if

        return ''.join(buffer)
    # end __str__

    def sumX(self) -> float:
        '''Determine the sum/total of all the numbers that have been given
           to this Statistician.

        Returns:
           The sum of all the numbers that have been given to this
           Statistician
        '''
        return self.__sumX
    # end sum

    @staticmethod
    def union(stat1:Statistician, stat2:Statistician) -> Statistician:
        '''Create and return a new Statistician that behaves as if it was
           given all the numbers from the two Statisticians stat1 and stat2.

        Preconditions:
           Both stat1 and stat2 must not be None and also must reference
           Statistician objects


Arguments:
            stat1  reference to a Statistician
            stat2  reference to a Statistician

        Returns:
            a new Statistician that behaves as if it was given all the
            numbers from the two Statisticians stat1 and stat2

        Raises:
            TypeError if either stat1 or stat2 is None or if either stat1 or
            stat2 not an instance of Statistician
        '''
        if stat1 is None or stat2 is None:
            raise TypeError("Either stat1 or stat2 is None")
        if not isinstance(stat1, Statistician) or\
           not isinstance(stat2, Statistician):
            raise TypeError("Either stat1 or stat2 is not a Statistician")

        temp = Statistician()

        # STUDENT completes implementation here
        return temp

# end class Statistician

File: CS2336/Lab01/Java/Handout/Statistician.java
// File: Statistician.java
// This is an assignment for students to complete after reading Chapter 2 of
// "Data Structures and Other Objects Using Java" by Michael Main.

import java.util.Objects;
import static java.lang.Math.sqrt;


/******************************************************************************
* This class is a homework assignment designed to help students begin
* their transition from Python into Java --
*
* A <CODE>Statistician</CODE> keeps track of statistics about a sequence of
* double numbers.<br>
*
* @version
*   September 2020  
******************************************************************************/
public class Statistician {   

   //  Here we define the "instance" variables that hold
   //  the "state" information for objects of this class.
   //  We must manage those "instance" variables appropriately
   //  since we are obligated to report to a user such "statistics" 
   //  as mean, sumX, largest, smallest, length and standard deviation. 
   //  Note that we need NOT save all the values
   //
   //  Class invariant:  the attributes/instance variables N, sumX,
   //  sumXsq, largest and smallest are used to derive the statistics
   //  and define the state of a statistician object

    private int    N;        // How many numbers given to statistician
    private double sumX;     // Sum of all the numbers given to statistician
    private double sumXsq;   // Sum of the squares of all numbers processed
    private double smallest; // The smallest number given to statistician
    private double largest;  // The largest number given to statistician


   /**
   * Initialize a new Statistician.
   * <b>Postcondition:</b><br>
   *   This Statistician is newly initialized and has not yet been given any 
   *   numbers.
   **/   
   public Statistician( ) {
      this.N        = 0;
      this.sumX     = 0.0;
      this.sumXsq   = 0.0;
      this.smallest = +Double.MAX_VALUE;
      this.largest  = -Double.MAX_VALUE;
   } 


   /**
   * Initialize a new Statistician as a copy of another Statistician.<br><br>
   * @param other
   *     other is a non-null reference to another Statisticiana<br><br>
   * <b>Precondition:</b><br>
   *   Other is not null.  If it is, then throw NullPointerException<br><br>
   * <b>Postcondition:</b><br>
   *   This Statistician is newly initialized and has been given attributes
   *   equal to those of other
   * @throws NullPointerException
   **/   
   public Statistician( Statistician other ) {
       if ( other == null )
           throw new NullPointerException("other cannot be null");

       this.N = other.N;
      // Student implementation -- initialize all remaining attributes
      // of "this" object with the values of the "other" object
   }  


   /**
   * Compare this <CODE>Statistician</CODE> to another object for equality.
   * @param obj
   *   an object with which this <CODE>Statistician</CODE> will be compared
   * @return
   *   A return value of <CODE>true</CODE> indicates that 
   *   <CODE>obj</CODE> refers to a 
   *   <CODE>Statistican</CODE> object with the same length, sum,
   *   sum of all numbers squared, smallest and largest as this 
   *   <CODE>Statistician</CODE>. Otherwise the return value is 
   *   <CODE>false</CODE>.<br>
   * <b>Note:</b><br>
   *   If <CODE>obj</CODE> is null or does not refer to a 
   *   <CODE>Statistician</CODE> object, then the return is 
   *   <CODE>false</CODE>.
   **/  
   @Override
   public boolean equals(Object obj) {
      if ( obj == null ) return false;
      if ( !(obj instanceof Statistician) ) return false;

      Statistician temp = (Statistician) obj;
      if ( this == temp ) return true; //Are we comparing with ourself
      if ( this.length( ) == 0 && temp.length( ) == 0 ) return true;
      // NOTE:  if two objects are equal, their hashCodes are equal
      if ( this.hashCode() != temp.hashCode() ) return false;

      // The student's code will remove the following return and replace
      // it with appropriate logic that compares each and every attribute
      // of this with attributes of temp
      
      return false;
   } 
   
   /**
    * Return a hashcode that is unique to an instance of
    * class Statistician based on the value of its 
    * attributes.
    * @return
    *   an int hash code that is unique to an object of
    *   class Statistician by combining the hash codes
    *   of all its attributes
    **/
   @Override
   public int hashCode() {
       return Objects.hash( this.N,
                            this.sumX,
                            this.sumXsq,
                            this.smallest,
                            this.largest 
                          );
   }
   
   /**
   * Determine how many numbers have been given to this Statistician.
   * @return
   *   the N of how many numbers have been given to this Statistician
   *   since it was initialized or reinitialized <br>
   * <b>Note:</b><br>
   *   Giving a Statistician more than
   *   <CODE>Integer.MAX_VALUE</CODE> numbers, will
   *   cause failure with an arithmetic overflow.
   **/ 
   public int length( ) {
      return this.N;
   }

   /**
   * Determine the largest number that has been given 
   * to this Statistician.
   * @return
   *   the largest number that has been given to this 
   *   Statistician since it was initialized or reinitialized<br>
   * <b>Note:</b><br>
   *   If <CODE>length()</CODE> is zero, then the return from this method
   *   is <CODE>Double.NaN</CODE>.
   **/ 
   public double largest( ) {
      // The student's code will replace the following return statement with
      // logic that matches specifications.
      return 0;
   }

   /**
   * Determine the arithmetic average of all the numbers that have been given 
   * to this Statistician.
   * @return
   *   the arithmetic mean of all the number that have been given to this 
   *   Statistician
   *   since it was initialized or reinitialized<br>
   * <b>Note:</b><br>
   *   If this Statistician has been given more than
   *   <CODE>Integer.MAX_VALUE</CODE> numbers, then this method fails
   *   because of arithmetic overflow.
   *   If <CODE>length()</CODE> is zero, then the return from this method
   *   is <CODE>Double.NaN</CODE>.
   *   If <CODE>sum()</CODE> exceeds the bounds of double numbers, then the 
   *   return from this method may be 
   *   <CODE>Double.POSITIVE_INFINITY</CODE> or
   *   <CODE>Double.NEGATIVE_INFINITY</CODE>.
   **/ 
   public double mean( ) {
      // The student's code will replace the following return statement with
      // code that matches specifications
      return 0;
   }


   /**
   * Give a new number to this Statistician. 
   * @param number
   *   the new number that is being given to this Statisticiana<br>
   * <b>Postcondition:</b><br>
   *   The specified number has been given to this Statistician and
   *   it will be included in any subsequent statistics
   **/
   public void nextNumber(double number) {
      // Student implementation -- must update all the
      // instance variables
      // NOTE:  if number is the first number given to the statistician,
      //        then number is both the largest and smallest seen so far.

   }
   
   /**
   * Reset this Statistician to state with no numbers
   * in the sequence. <br>
   *
   * <b>Postcondition:</b><br>
   *   This Statistician is reinitialized as if it has never been given any 
   *   numbers
   **/
   public void reset( ) {
      // Student implementation.
   }

   /**
    * Create a new Statistician whose state is scaled by parameter scale--
    * such that "statistics" other than length() are multiplied by scale.
    * <b>Postcondition:</b><br>
    *   The returned Statistician is a scaled copy of parameter s which 
    *   is not modified.
    * @param scale
    *   a double scale factor
    * @param s
    *   a reference to a Statistician
    * @return
    *   a reference to a new Statiscian that is a scaled copy of parameter s
    **/
   public static Statistician scale(double scale, Statistician s) {
       Statistician temp = new Statistician(s);

      // Student implementation follows
      
       return temp;
   }

   /**
   * Determine the smallest number that has been given 
   * to this Statistician.
   * @return
   *   the smallest number that has been given to this 
   *   Statistician since it was initialized or reinitialized<br>
   * <b>Note:</b><br>
   *   If <CODE>length()</CODE> is zero, then the return from this method
   *   is <CODE>Double.NaN</CODE>.
   **/ 
   public double smallest( ) {
      // The student's code will replace the following return statement with
      // code that matches specifications
      return 0;
   }

   /**
   * Determine the standard deviation of the stream of numbers
   * given to this statistician.
   * <b>Postcondition:</b><br>
   *   State of this statistician is NOT altered
   *
   * @return
   *    The standard deviation of the stream of numbers given
   *    to this statistician -- stream of numbers treated as a
   *    population.
   * <b>Note:</b><br>
   *   If <CODE>length()</CODE> is zero, then the return from this method
   *   is <CODE>Double.NaN</CODE>.
   **/
   public double stdDev() {
       if ( this.length() == 0 )
           return Double.NaN;
       //Student implementation replace the following return with
       //appropriate computations and return
       return 0.0;
   }

   /**
   * Create a string representation of this Statistician.<br><br>
   * <b>Postcondition:</b><br>
   *   State of the Statistician remains unaltered by this method
   * @return String
   *   a human readable representation of the Statistician's internal
   *   state
   **/
   @Override
   public String toString( ) {
       StringBuffer sb = new StringBuffer();
       sb.append("Statistician: \n");
       sb.append("\tN      = " + this.N + "\n");
       sb.append("\tsumX   = " + this.sumX + "\n");
       sb.append("\tsumXsq = " + this.sumXsq + "\n");
       if ( this.N == 0 ) {
           sb.append("\tmean       = NOT DEFINED\n");
           sb.append("\tstd. dev   = NOT DEFINED\n");
           sb.append("\tsmallest   = NOT DEFINED\n");
           sb.append("\tlargest    = NOT DEFINED\n");
       }
       else {
           sb.append("\tmean       = " + this.mean()  + "\n");
           sb.append("\tstd. dev   = " + this.stdDev() + "\n");
           sb.append("\tsmallest   = " + this.smallest + "\n");
           sb.append("\tlargest    = " + this.largest + "\n");
       }
       return sb.toString();
   }

   /**
   * Determine the sum of all the numbers that have been given to this 
   * Statistician.
   * @return
   *   the sum of all the number that have been given to this 
   *   Statistician since it was initialized or reinitialized <br>
   * <b>Note:</b><br>
   *   If the sum exceeds the bounds of double numbers, then the answer
   *   from this method may be 
   *   <CODE>Double.POSITIVE_INFINITY</CODE> or
   *   <CODE>Double.NEGATIVE_INFINITY</CODE>.
   **/ 
   public double sumX( ) {
      // The student's code will replace the following return with the 
      // appropriate return
      return 0;
   }

   /**
   * Create a new Statistician that behaves as if it were given all the
   * numbers from two other statisticians.
   * @param  s1
   *   the first of two Statisticians
   * @param s2
   *   the second of two Statisticians <br>
   * <b>Precondition:</b><br>
   *   Neither s1 nor s2 is null.
   * @return
   *   a new Statistician that acts as if it were given all the numbers from
   *   both s1 and s2.
   * @throws NullPointerException
   *   Indicates that one or the other or both of the arguments are null.
   **/   
   public static Statistician union(Statistician s1, Statistician s2) {
      if ( s1 == null || s2 == null )
          throw new NullPointerException("One or both arguments are null");

      Statistician temp = new Statistician();
      // Student's code to correctly set temp to be union

      return temp;
   }
      
}
           
